src/cargo/sources/registry/remote.rs

```diff
@@ -176,6 +177,14 @@ impl<'cfg> RegistryData for RemoteRegistry<'cfg> {
 
     fn update_index(&mut self) -> CargoResult<()> {
         if self.config.cli_unstable().offline {
+            if self.repo()?.is_empty()? {
+                failure::bail!(
+                    "unable to fetch {} in offline mode\n\
+                     Run `cargo fetch` to download the registry index and all \
+                     of a project's dependencies before going offline.",
+                    self.source_id
+                );
```

Comment:
Could you add a comment here that we're explicitly switching the `Ok` to an `Err` because if we're updating the index we're guaranteed to require at leats one crate and since the index is empty it's guaranteed to have zero crates, therefore we're trying to give a better error message

Reply:
Hm, this suggestion is also a little sketchy.  If you run `cargo install foo` for the first time, and the index is unavailable, it suggests running `cargo fetch` which will definitely not work.  Maybe it is not worth it?

Reply:
That's a good point yeah, but I think that could be fixed by tweaking the error message? It could mention that you should try running with out `--offline` or otherwise run `cargo fetch` sooner perhaps?